In [1]:
import pandas as pd
import os

In [2]:
title_basic = pd.read_csv('db_files/title_basic.tsv', delimiter='\t')

C:\Users\Lorenzo\AppData\Local\Temp\ipykernel_7000\3681487345.py:1: DtypeWarning: Columns (4) have mixed types. Specify dtype option on import or set low_memory=False.
  title_basic = pd.read_csv('db_files/title_basic.tsv', delimiter='\t')


In [3]:
title_basic.shape

(9647020, 9)

In [4]:
title_basic.head(20)

tconst titleType                                 primaryTitle  \
0   tt0000001     short                                   Carmencita   
1   tt0000002     short                       Le clown et ses chiens   
2   tt0000003     short                               Pauvre Pierrot   
3   tt0000004     short                                  Un bon bock   
4   tt0000005     short                             Blacksmith Scene   
5   tt0000006     short                            Chinese Opium Den   
6   tt0000007     short  Corbett and Courtney Before the Kinetograph   
7   tt0000008     short       Edison Kinetoscopic Record of a Sneeze   
8   tt0000009     movie                                   Miss Jerry   
9   tt0000010     short                          Leaving the Factory   
10  tt0000011     short                      Akrobatisches Potpourri   
11  tt0000012     short                       The Arrival of a Train   
12  tt0000013     short  The Photographical Congress Arrives in Lyon   
13  tt0000014     short                          The Waterer Watered   
14  tt0000015     short                          Autour d'une cabine   
15  tt0000016     short                        Boat Leaving the Port   
16  tt0000017     short                     Italienischer Bauerntanz   
17  tt0000018     short                         Das boxende Känguruh   
18  tt0000019     short                             The Clown Barber   
19  tt0000020     short                               The Derby 1895   

                                        originalTitle isAdult startYear  \
0                                          Carmencita       0      1894   
1                              Le clown et ses chiens       0      1892   
2                                      Pauvre Pierrot       0      1892   
3                                         Un bon bock       0      1892   
4                                    Blacksmith Scene       0      1893   
5                                   Chinese Opium Den       0      1894   
6         Corbett and Courtney Before the Kinetograph       0      1894   
7              Edison Kinetoscopic Record of a Sneeze       0      1894   
8                                          Miss Jerry       0      1894   
9                 La sortie de l'usine Lumière à Lyon       0      1895   
10                            Akrobatisches Potpourri       0      1895   
11                   L'arrivée d'un train à La Ciotat       0      1896   
12  Le débarquement du congrès de photographie à Lyon       0      1895   
13                                  L'arroseur arrosé       0      1895   
14                                Autour d'une cabine       0      1894   
15                             Barque sortant du port       0      1895   
16                           Italienischer Bauerntanz       0      1895   
17                               Das boxende Känguruh       0      1895   
18                                   The Clown Barber       0      1898   
19                                     The Derby 1895       0      1895   

   endYear runtimeMinutes                    genres  
0       \N              1         Documentary,Short  
1       \N              5           Animation,Short  
2       \N              4  Animation,Comedy,Romance  
3       \N             12           Animation,Short  
4       \N              1              Comedy,Short  
5       \N              1                     Short  
6       \N              1               Short,Sport  
7       \N              1         Documentary,Short  
8       \N             45                   Romance  
9       \N              1         Documentary,Short  
10      \N              1         Documentary,Short  
11      \N              1         Documentary,Short  
12      \N              1         Documentary,Short  
13      \N              1              Comedy,Short  
14      \N              2           Animation,Short  
15      \N              1         Documentary,Short  


## Filtering steps:

1. titleType:
    - movie
    - tvSeries
    - tvMovie

2. startYear: year >= 1980
3. num of replicas: >= 1
4. num of reviews : >= 1000
5. num of actor per movie: >= 1


In [5]:
allowed = ['movie', 'tvSeries', 'tvMovie']
startYear = 1980
n_replicas = 1
n_reviews = 1000
null_values = ['\\N', '\\n', '']

In [6]:
title_basic['startYear'] = title_basic['startYear'].replace('\\N', 0).astype('int32')
title_basic['endYear'] = title_basic.apply(
    lambda r: int(r['startYear']) if r['endYear'] == '\\N' else int(r['endYear']), axis=1
)
title_basic['runtimeMinutes'] = title_basic['runtimeMinutes'].replace('\\N', 0)
title_basic = title_basic.drop(title_basic[title_basic['genres'].isin(null_values)].index)

In [7]:
title_basic['genres'] = title_basic['genres'].str.split(',')

In [8]:
g = title_basic['genres']
g = g.explode()
g.value_counts()

Drama           2725059
Comedy          1959541
Talk-Show       1162771
Short           1059508
Documentary      912034
Romance          889406
News             843408
Family           709851
Reality-TV       536574
Animation        495894
Crime            396347
Action           384377
Adventure        378884
Music            376819
Game-Show        353205
Adult            296861
Sport            228226
Fantasy          198500
Mystery          188956
Horror           168532
Thriller         161884
History          143369
Sci-Fi           106342
Biography        105568
Musical           86713
War               33230
Western           29398
Film-Noir           890
Experimental          1
Name: genres, dtype: int64

+------------+-------+
|            |genres |
+------------+-------+
|Drama       |2725059|
|Comedy      |1959541|
|Talk-Show   |1162771|
|Short       |1059508|
|Documentary |912034 |
|Romance     |889406 |
|News        |843408 |
|Family      |709851 |
|Reality-TV  |536574 |
|Animation   |495894 |
|Crime       |396347 |
|Action      |384377 |
|Adventure   |378884 |
|Music       |376819 |
|Game-Show   |353205 |
|Adult       |296861 |
|Sport       |228226 |
|Fantasy     |198500 |
|Mystery     |188956 |
|Horror      |168532 |
|Thriller    |161884 |
|History     |143369 |
|Sci-Fi      |106342 |
|Biography   |105568 |
|Musical     |86713  |
|War         |33230  |
|Western     |29398  |
|Film-Noir   |890    |
|Experimental|1      |
+------------+-------+


In [9]:
m1 = title_basic['startYear'] >= startYear
m2 = title_basic['titleType'].isin(allowed)
db = title_basic[m1 & m2]
del title_basic, g
db.head(20)

tconst titleType                                       primaryTitle  \
11636  tt0011801     movie                                   Tötet nicht mehr   
13082  tt0013274     movie                        Istoriya grazhdanskoy voyny   
15486  tt0015724     movie                                      Dama de noche   
34803  tt0035423     movie                                     Kate & Leopold   
35967  tt0036606     movie                        Another Time, Another Place   
37419  tt0038086     movie                          Shiva und die Galgenblume   
38011  tt0038687     movie                                 Let There Be Light   
53242  tt0054261   tvMovie                         Rómeó, Júlia és a sötétség   
56372  tt0057461     movie                             La rosa de los vientos   
58202  tt0059325     movie                                        Born in '45   
58756  tt0059900     movie                     Wenn du groß bist, lieber Adam   
59684  tt0060854   tvMovie                                 Preis der Feigheit   
60970  tt0062181     movie                                       Rece do góry   
61116  tt0062336     movie  The Tango of the Widower and Its Distorting Mi...   
61389  tt0062614  tvSeries                                        Zeg 'ns Aaa   
61612  tt0062842   tvMovie                                     Cuore di mamma   
63536  tt0064820     movie                             The Plot Against Harry   
63699  tt0064994     movie                                  Larks on a String   
63811  tt0065108     movie                                    Tolko tri nochi   
63890  tt0065188     movie                            Vojtech, receny sirotek   

                                   originalTitle isAdult  startYear  endYear  \
11636                           Tötet nicht mehr       0       2019     2019   
13082                Istoriya grazhdanskoy voyny       0       2021     2021   
15486                              Dama de noche       0       1993     1993   
34803                             Kate & Leopold       0       2001     2001   
35967                Another Time, Another Place       0       1983     1983   
37419                  Shiva und die Galgenblume       0       1993     1993   
38011                         Let There Be Light       0       1980     1980   
53242                 Rómeó, Júlia és a sötétség       0       1986     1986   
56372                     La rosa de los vientos       0       1983     1983   
58202                                Jahrgang 45       0       1990     1990   
58756             Wenn du groß bist, lieber Adam       0       1990     1990   
59684                         Preis der Feigheit       0       1986     1986   
60970                               Rece do góry       0       1981     1981   
61116  El tango del viudo y su espejo deformante       0       2020     2020   
61389                                Zeg 'ns Aaa       0       1981     2010   
61612                             Cuore di mamma       0       1988     1988   
63536                     The Plot Against Harry       0       1989     1989   
63699                          Skrivánci na niti       0       1990     1990   
63811                            Tolko tri nochi       0       1989     1989   
63890                    Vojtech, receny sirotek       0       1990     1990   

      runtimeMinutes                      genres  
11636              0             [Action, Crime]  
13082            133               [Documentary]  
15486            102   [Drama, Mystery, Romance]  
34803            118  [Comedy, Fantasy, Romance]  
35967            118                [Drama, War]  
37419              0                  [Thriller]  
38011             58          [Documentary, War]  
53242            101                     [Drama]  
56372             84            [Drama, Fantasy]  
58202            100            [Drama, Romance]  
58756             78            [Drama, Fantasy]  
59684          

In [10]:
db.shape

(639084, 9)

In [11]:
title_akas = pd.read_csv('db_files/title_akas.tsv', delimiter='\t')
title_akas = title_akas[['titleId','ordering','region']]

C:\Users\Lorenzo\AppData\Local\Temp\ipykernel_7000\1726883815.py:1: DtypeWarning: Columns (7) have mixed types. Specify dtype option on import or set low_memory=False.
  title_akas = pd.read_csv('db_files/title_akas.tsv', delimiter='\t')


In [12]:
title_akas = title_akas.drop(title_akas[title_akas['region'].isin(null_values)].index)
title_akas = title_akas.drop(title_akas[title_akas['ordering'].isin(null_values)].index)
title_akas = title_akas[title_akas['titleId'].isin(db.tconst)]

In [13]:
title_akas.shape

(2135913, 3)

In [14]:
replicas = title_akas.groupby('titleId').aggregate({'ordering':'count', 'titleId':'first'})
replicas = replicas[replicas['ordering'] >= n_replicas]
title_akas = title_akas[title_akas['titleId'].isin(replicas.titleId)]
title_akas.shape

(2135913, 3)

Regions

In [15]:
regions = title_akas.groupby('titleId').aggregate(
    {'ordering': 'count', 'region': lambda e: list(e)}
).sort_values('ordering', ascending=False)
regions = regions.rename(columns={'ordering':'replicas', 'region':'regions'})
regions

replicas                                            regions
titleId                                                                
tt0088814        247  [DZ, FI, IT, BE, CG, RS, CN, CR, SE, AT, CZ, D...
tt0168366        189  [US, JP, AR, CN, SI, TW, US, US, SE, US, IN, U...
tt2872750        149  [ES, XWW, EG, GB, CM, PE, AR, PR, LB, US, QA, ...
tt0099785        144  [PL, KR, BD, MA, UZ, IN, MD, PL, FR, PT, LV, S...
tt0104431        134  [MD, UA, HR, KR, FR, PL, EC, SG, FI, SE, VN, A...
...              ...                                                ...
tt2047917          1                                               [US]
tt10303906         1                                               [CY]
tt10303908         1                                               [CL]
tt2047907          1                                               [US]
tt9916754          1                                               [BR]

[633844 rows x 2 columns]

In [16]:
db = db.join(regions, on='tconst').dropna()
del title_akas
del regions

In [17]:
name_basics = pd.read_csv('db_files/name_basics.tsv', delimiter='\t')

In [18]:
name_basics = name_basics.drop(name_basics[name_basics['birthYear'] == '\\N'].index)
name_basics['deathYear'] = name_basics['deathYear'].replace({'\\N':'2024'})
actors = name_basics[
    name_basics['primaryProfession'].str.contains('actor') |
    name_basics['primaryProfession'].str.contains('actress')
]
actors = actors.drop(actors[actors['knownForTitles'] == '\\N'].index)

In [19]:
actors['knownForTitles'] = actors['knownForTitles'].apply(lambda e: e.split(',')).explode('knownForTitles')
del name_basics

In [20]:
actors = actors[actors['knownForTitles'].isin(db.tconst)]
actors = actors.rename(columns={'knownForTitles':'tconst'})
db = db.join(actors.set_index('tconst'), on='tconst').dropna()
del actors
db.head(20)

tconst titleType                 primaryTitle  \
15486  tt0015724     movie                Dama de noche   
34803  tt0035423     movie               Kate & Leopold   
34803  tt0035423     movie               Kate & Leopold   
34803  tt0035423     movie               Kate & Leopold   
34803  tt0035423     movie               Kate & Leopold   
34803  tt0035423     movie               Kate & Leopold   
35967  tt0036606     movie  Another Time, Another Place   
58202  tt0059325     movie                  Born in '45   
60970  tt0062181     movie                 Rece do góry   
61389  tt0062614  tvSeries                  Zeg 'ns Aaa   
61389  tt0062614  tvSeries                  Zeg 'ns Aaa   
61389  tt0062614  tvSeries                  Zeg 'ns Aaa   
63536  tt0064820     movie       The Plot Against Harry   
63536  tt0064820     movie       The Plot Against Harry   
63699  tt0064994     movie            Larks on a String   
63699  tt0064994     movie            Larks on a String   
63699  tt0064994     movie            Larks on a String   
63699  tt0064994     movie            Larks on a String   
63811  tt0065108     movie              Tolko tri nochi   
63890  tt0065188     movie      Vojtech, receny sirotek   

                     originalTitle isAdult  startYear  endYear runtimeMinutes  \
15486                Dama de noche       0       1993     1993            102   
34803               Kate & Leopold       0       2001     2001            118   
34803               Kate & Leopold       0       2001     2001            118   
34803               Kate & Leopold       0       2001     2001            118   
34803               Kate & Leopold       0       2001     2001            118   
34803               Kate & Leopold       0       2001     2001            118   
35967  Another Time, Another Place       0       1983     1983            118   
58202                  Jahrgang 45       0       1990     1990            100   
60970                 Rece do góry       0       1981     1981             76   
61389                  Zeg 'ns Aaa       0       1981     2010             23   
61389                  Zeg 'ns Aaa       0       1981     2010             23   
61389                  Zeg 'ns Aaa       0       1981     2010             23   
63536       The Plot Against Harry       0       1989     1989             81   
63536       The Plot Against Harry       0       1989     1989             81   
63699            Skrivánci na niti       0       1990     1990             94   
63699            Skrivánci na niti       0       1990     1990             94   
63699            Skrivánci na niti       0       1990     1990             94   
63699            Skrivánci na niti       0       1990     1990             94   
63811              Tolko tri nochi       0       1989     1989             82   
63890      Vojtech, receny sirotek       0       1990     1990             80   

                           genres  replicas  \
15486   [Drama, Mystery, Romance]       2.0   
34803  [Comedy, Fantasy, Romance]      41.0   
34803  [Comedy, Fantasy, Romance]      41.0   
34803  [Comedy, Fantasy, Romance]      41.0   
34803  [Comedy, Fantasy, Romance]      41.0   
34803  [Comedy, Fantasy, Romance]      41.0   
35967                [Drama, War]      11.0   
58202            [Drama, Romance]       7.0   
60970                     [Drama]      10.0   
61389                    [Comedy]       4.0   
61389                    [Comedy]       4.0   
61389                    [Comedy]       4.0   
63536             [Comedy, Crime]       7.0   
63536             [Comedy, Crime]       7.0   
63699    [Comedy, Drama, Romance]      27.0   
63699    [Comedy, Drama, Romance]      27.0   
63699    [Comedy, Drama, Romance]      27.0   
63699    [Comedy, Drama, Romance]      27.0   
63811            [Drama, Romance]       1.0   
63890                     [Drama]       1.0   

                                                 regions     nconst  \
15486    

In [21]:
db.shape

(91893, 16)

In [22]:
title_ratings = pd.read_csv('db_files/title_ratings.tsv', delimiter='\t')

In [23]:
title_ratings = title_ratings.drop(title_ratings[title_ratings['averageRating'] == '\\N'].index)
title_ratings = title_ratings.drop(title_ratings[title_ratings['numVotes'] == '\\N'].index)
title_ratings = title_ratings[title_ratings['numVotes'] >= n_reviews]
title_ratings = title_ratings[title_ratings['tconst'].isin(db.tconst)]

In [24]:
db = db.merge(title_ratings, on='tconst').dropna()
del title_ratings
db.shape

(42874, 18)

Store the cleaned database

In [25]:
import numpy as np
for index in db.index:
    for column in db.columns:
        cell_entry = db.loc[index, column]
        if (type(cell_entry) == str) | (type(cell_entry) == np.str_) :
            db.loc[index, column] = cell_entry.replace('"', '')
            db.loc[index, column] = cell_entry.replace('\'', '')

In [26]:
# db = ...
os.makedirs('db_clean', exist_ok=True)
db.to_csv('db_clean/titles.csv', index=False)